test: Unit test cases for Systemd Watchdog

tvyas opened the pull request:
Add unit test cases to the existing Google test framework for the AVB StreamHandler Systemd Watchdog.

keerockl commented:
Looks good to me.

tvyas commented:
This has caught a few bugs, or more like while writing these unit tests, I found a few bugs in the watchdog code. Once we merge this, I'll submit the fixes for those bugs.

edersondisouza commented:
Why not sending the fixes now?

tvyas commented:
> Why not sending the fixes now?
I was like it'd be good to have some of the tests fail, and then send the fixes and have them pass, but I suppose I can send the fixes too. Any which way is fine.

edersondisouza commented:
Tests failing can be an issue when using git bisect - so let's have all
on this PR, and with the fixes before the tests =D

On Thu, 2018-11-01 at 16:30 -0700, tvyas wrote:
> > Why not sending the fixes now?
> > I was like it'd be good to have some of the tests fail, and then send the fixes and have them pass, but I suppose I can send the fixes too. Any which way is fine.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

tvyas commented:
> Tests failing can be an issue when using git bisect - so let's have all on this PR, and with the fixes before the tests =D
> […](#)
> On Thu, 2018-11-01 at 16:30 -0700, tvyas wrote: > Why not sending the fixes now? > I was like it'd be good to have some of the tests fail, and then send the fixes and have them pass, but I suppose I can send the fixes too. Any which way is fine. — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or mute the thread.

Done

edersondisouza commented:
Oh, I wasn't clear, sorry - but I meant to have more than one commit on this PR - the commits that fixes the bugs, and the final commit, with the tests. Having them all on the same commit ends up being messy...

edersondisouza commented on private/src/avb_watchdog/IasSystemdWatchdogManager.cpp:
```diff
@@ -144,8 +145,11 @@ IasWatchdogInterface* IasSystemdWatchdogManager::createWatchdog()
 IasWatchdogInterface* IasSystemdWatchdogManager::createWatchdog(uint32_t timeout, std::string watchdogName)
 {
   IasWatchdogInterface * watchdogInterface = nullptr;
+  int signed_timeout = timeout;
```
This is weird... is the idea prevent a big timeout? Why not compare with threshold then? Commit message mentions "assigning negative timeout value to an unsigned int" - but that should be fixed at the caller, not here.

tvyas replied:
Ah, yea, it'd be better to directly compare with the threshold. 

IMO, the function should validate the unsigned int irrespective of what the caller is doing. Practically, the caller will never pass a negative value to an unsigned int, but still, the function should not assume that and do the required validation of the inputs. What do you think ?

keerockl replied:
the value could be converted into negative value when it's assigned to int if timeout value is big enough.

keerockl replied:
never mind, already answered to the question.

edersondisouza commented on private/src/avb_watchdog/IasWatchdogInterface.cpp:
```diff
@@ -46,23 +46,15 @@ IasWatchdogInterface::~IasWatchdogInterface()
 IasWatchdogResult IasWatchdogInterface::registerWatchdog(uint32_t timeout, std::string watchdogName)
 {
   IasWatchdogResult result = IasMediaTransportAvb::IasResult::cOk;
+  int signed_timeout = timeout;
```
Ditto. Where is it possible for someone to send a negative value here? The API is quite clear about `uint32_t`.

tvyas replied:
Yes, but as I said, is it safe to rely solely on the callers doing the *right* thing. We should still validate  what's being given to us.

edersondisouza reviewed (commented):
Didn't review the test itself yet, just the fixes. Will review more later.

edersondisouza commented:
On Sat, 2018-11-03 at 12:22 -0700, tvyas wrote:
> @tvyas commented on this pull request.
> In private/src/avb_watchdog/IasWatchdogInterface.cpp:
> > @@ -46,23 +46,15 @@ IasWatchdogInterface::~IasWatchdogInterface()
>  IasWatchdogResult IasWatchdogInterface::registerWatchdog(uint32_t timeout, std::string watchdogName)
>  {
>    IasWatchdogResult result = IasMediaTransportAvb::IasResult::cOk;
> +  int signed_timeout = timeout;
> Yes, but as I said, is it safe to rely solely on the callers doing the right thing. We should still validate what's being given to us.
> —

I don't think we should worry about this. I mean, what if instead of
2^31 + 1 (a negative value), caller sends 2^31 - 1 (a positive value),
but still, really, really big? What would the threshold be?

To prevent mistakes, maybe use the `-Wsign-conversion` warning?

> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

tvyas committed:
tests: Add unit test cases for Systemd watchdog

Signed-off-by: Tarun Vyas <tarun.vyas@intel.com>

tvyas committed:
watchdog: Bunch of minor fixes for watchdog

Signed-off-by: Tarun Vyas <tarun.vyas@intel.com>

tvyas commented:
> On Sat, 2018-11-03 at 12:22 -0700, tvyas wrote: @tvyas commented on this pull request. In private/src/avb_watchdog/IasWatchdogInterface.cpp: > @@ -46,23 +46,15 @@ IasWatchdogInterface::~IasWatchdogInterface() IasWatchdogResult IasWatchdogInterface::registerWatchdog(uint32_t timeout, std::string watchdogName) { IasWatchdogResult result = IasMediaTransportAvb::IasResult::cOk; + int signed_timeout = timeout; Yes, but as I said, is it safe to rely solely on the callers doing the right thing. We should still validate what's being given to us. —
> I don't think we should worry about this. I mean, what if instead of 2^31 + 1 (a negative value), caller sends 2^31 - 1 (a positive value), but still, really, really big? What would the threshold be? To prevent mistakes, maybe use the `-Wsign-conversion` warning?
Sounds like an idea !
> […](#)
> You are receiving this because you commented. Reply to this email directly, view it on GitHub, or mute the thread.

tvyas commented:
Split the original commit into 2 and addressed other review comments.

edersondisouza commented on private/tst/avb_streamhandler/CMakeLists.txt:
```diff
@@ -33,9 +33,10 @@ add_executable( test_IasTestAvbStreamhandler
                 )
 
-target_compile_options( test_IasTestAvbStreamhandler PRIVATE -Wno-error )
+target_compile_options( test_IasTestAvbStreamhandler PRIVATE -Wno-error -Wsign-conversion)
```
Please, add an issue so we can track the task of adding this to avb-sh build itself.

edersondisouza commented on private/tst/avb_streamhandler/src/IasTestSystemdWatchdog.cpp:
```diff
@@ -0,0 +1,233 @@
+
+    mWatchdogTimerRegistration = std::make_shared<IasWatchdog::IasWatchdogTimerRegistration>(mDltContext);
+    ASSERT_TRUE(mWatchdogTimerRegistration != nullptr);
+
```
Astray extra line here.

edersondisouza commented on private/tst/avb_streamhandler/src/IasTestSystemdWatchdog.cpp:
```diff
@@ -0,0 +1,233 @@
+{
+  ASSERT_EQ(setenv("WATCHDOG_USEC", "30000000", 1), 0);
+  IasWatchdogResult result = mWdManager->init(mWatchdogTimerRegistration);
+  ASSERT_EQ(result, IasMediaTransportAvb::IasResult::cOk);
```
Maybe use `EXPECT_EQ` here? This way we can keep on testing other scenarios. Note that the `ASSERT_EQ` on the `setenv` call is ok, as that call is not what we are testing, and if it fails, it compromises the result of the test.

edersondisouza commented on private/tst/avb_streamhandler/src/IasTestSystemdWatchdog.cpp:
```diff
@@ -0,0 +1,233 @@
+  IasWatchdogInterface *wdIf = nullptr;
+  IasWatchdogResult res = IasMediaTransportAvb::IasResult::cOk;
+
+  wdIf = mWdManager->createWatchdog(30000, "Ultron");
```
Nice! But maybe this name is registered, not sure if we can use it here... =/

edersondisouza commented on private/tst/avb_streamhandler/src/IasTestSystemdWatchdog.cpp:
```diff
@@ -0,0 +1,233 @@
+  ASSERT_EQ(res, IasMediaTransportAvb::IasResult::cObjectNotFound);
+
+  res = mWdManager->removeWatchdog(nullptr);
+  ASSERT_EQ(res, IasMediaTransportAvb::IasResult::cParameterInvalid);
```
Note that all assertions for the `removeWatchdog` results could be `EXPECT_XX`. The one that creates the watchdog should remain an assertion, though.

edersondisouza commented on private/tst/avb_streamhandler/src/IasTestSystemdWatchdog.cpp:
```diff
@@ -0,0 +1,233 @@
+  ASSERT_EQ(res, IasWatchdogResult::cWatchdogNotPreconfigured);
+
+  wdIf->setTimeout(30000);
+  wdIf->setName("Thanos");
```
Ditto.

edersondisouza commented on private/tst/avb_streamhandler/src/IasTestSystemdWatchdog.cpp:
```diff
@@ -0,0 +1,233 @@
+  IasWatchdogInterface *wdIf = nullptr;
+  IasWatchdogResult res = IasMediaTransportAvb::IasResult::cOk;
+
+  wdIf = mWdManager->createWatchdog(30000, "Loki");
```
This one should be ok, though =P

edersondisouza commented on private/tst/avb_streamhandler/src/IasTestSystemdWatchdog.cpp:
```diff
@@ -0,0 +1,233 @@
+  res = wdIf->reset();
+  ASSERT_EQ(res, IasMediaTransportAvb::IasResult::cOk);
+
+  sleep(35);
```
Can't this time be smaller? Waiting 35 seconds appears... to much.

edersondisouza commented on private/tst/avb_streamhandler/src/IasTestSystemdWatchdog.cpp:
```diff
@@ -0,0 +1,233 @@
+  wdIf_1 = mWdManager->createWatchdog(30000, "Hela");
+  wdIf_2 = mWdManager->createWatchdog(30000, "Surtur");
+
+  ASSERT_FALSE(allValid);
```
I don't see this point of this assertion...

edersondisouza reviewed (changes requested):
Looking ok. But please see the opportunities to use `EXPECT_XX` instead of `ASSERT_XX`, as that would allow more tests to be run. `ASSERT_XX` makes sense only when continuing running the tests doesn't make sense, like when some preparation method fails.

tvyas committed:
watchdog: Promote the watchdog timeout calculation to a ULL

Currently, the watchdog timeout is calculated as
(mTimeout * 1000000) which causes an issue with larger
mTimeout values b/c mTimeout (in msec) is declared as
a uint32_t and the conversion to nanosec above will fail
miserably.

Modify this to (mTimeout * 1000000ULL) to promote the
operation to a uint64_t so that larger mTimeout values
can be accomodated

Signed-off-by: Tarun Vyas <tarun.vyas@intel.com>

edersondisouza commented on private/tst/avb_streamhandler/src/IasTestSystemdWatchdog.cpp:
```diff
@@ -0,0 +1,230 @@
+
+  sleep(10);
+  res = wdIf->reset();
+  EXPECT_EQ(res, IasWatchdogResult::cTimedOut);
```
Maybe I understood it wrong here, but when you created this watchdog with `30000`, it would take 30 seconds to fire, right? Then you sleep for like 10 seconds, and expects it to fail due time out. Will this work? Shouldn't you create the watchdog with a smaller time, like 3 seconds? Then it would have timed out on this test, after 10 seconds waiting. Unless, as I said, I understood things wrong...

tvyas replied:
Ok, seerious and shtuupid bug has been caught. The reset() function does the following:

       _mDueResetTime =  currentTime + (mTimeout * 1000000);_

where mTimeout is what we pass, 30 seconds in this case.

Now, mTimeout has been declared as a uint32_t. For large values like 30 seconds = (30,000 * 1,000,000) nanoseconds, this won't work, obviously. 
I'll include another patch to change mTimeout to a uint64_t. Will reduce the timer here from 30 sec to 5 sec too.

edersondisouza replied:
Ah, the value of testing =D
Yeah, if we keep time in nanoseconds, `uint64_t` is mandatory. But do we need to keep time in nanoseconds?

tvyas replied:
The comparison against current_time (another uint64_t) won't work then. It's being derived directly from clock_gettime() and the nanosecond component is also being considered there. Also, other parameters such as _dueResetTime_ are also declared as uint64_t. So, changing mTimeout's type looks like the fastest way out, right now.

tvyas replied:
Functionally, I agree, we don't need a nanosecond accuracy here.

edersondisouza replied:
Fair enough.

tvyas replied:
Done. Thank for the review :)

edersondisouza commented on private/tst/avb_streamhandler/src/IasTestSystemdWatchdog.cpp:
```diff
@@ -0,0 +1,230 @@
+  EXPECT_EQ(res, IasMediaTransportAvb::IasResult::cOk);
+
+  res = wdIf->registerWatchdog(30000, "Sif");
+  ASSERT_EQ(res, IasWatchdogResult::cAlreadyRegistered);
```
Nitpick: `EXPECT_EQ` here?

tvyas replied:
Well, this was the end of the test, so I was like it won't matter...

edersondisouza replied:
True that, but if someday more tests are added below... is one less thing to worry about.

tvyas replied:
Agreed. Modified in the next rev.

tvyas replied:
Done

edersondisouza reviewed (commented):
Looking ok, just some comment about something I'm not sure I understood.

edersondisouza reviewed (commented):
Oh... if `mTimeout` contains milliseconds, maybe all you should do is to change the multiplication to something like `(uint64_t)mTimeout * 1000000ULL`, no? For a while I thought that `mTimeout` had the nanoseconds and it was only working due small number of seconds.

tvyas commented:
Doh, I did think of an easier way, but just went ahead with the impulse to promote the type to uint64_t. 
Shouldn't _mTimeout * 1000000ULL_ suffice though ? The ULL should promote the whole thing.

edersondisouza commented:
> Doh, I did think of an easier way, but just went ahead with the impulse to promote the type to uint64_t.
> Shouldn't _mTimeout * 1000000ULL_ suffice though ? The ULL should promote the whole thing.

Yep. (I just like being more explicit sometimes, but you're right =D )

tvyas commented:
> > Doh, I did think of an easier way, but just went ahead with the impulse to promote the type to uint64_t.
> > Shouldn't _mTimeout * 1000000ULL_ suffice though ? The ULL should promote the whole thing.
> 
> Yep. (I just like being more explicit sometimes, but you're right =D )

Thanks again. Done !

edersondisouza reviewed (approved):
:+1:
